docs/unittesting.md

```diff
@@ -0,0 +1,264 @@
+# Unit Testing in TorQ
+
+### Introduction
+
+Unit testing is a software validation methodology where individual modules or 'units' of source code are tested to determine whether they work as intended. TorQ's unit testing framework has been designed to be user-friendly and includes lots of useful features such as debugging and integration testing. We shall see that, using this framework, it is straightforward to test both small units of code as well as larger pieces of application functionality.
+
+### Test Basics
+
+**My First Test**
+
+Tests are written in the style of Kx's [k4unit](https://code.kx.com/q/kb/unit-tests/), where tests are written in CSV files and loaded into a Q process along with the testing code and are run line by line, with the results being stored in a table. In order to write and run a basic test, in a new folder create a CSV file with some basic tests in it, like the following:
+
+```
+action,ms,bytes,lang,code,repeat,minver,comment
+comment,0,0,,this will be ignored,1,,""
+before,0,0,q,aa:22,1,,"This sets a variable before the tests begin"
+true,0,0,q,2=sum 1 1,1,,"A basic arithmetic check"
+fail,0,0,q,2=`aa,1,,"This ought to fail"
+after,0,0,q,bb:33,1,,"This code executes after the tests have run"
+```
+
+Then use the `torq.q` script to start up a TorQ process with the test flag pointing to the directory where the CSV is written and a debug flag so that the process outputs its logs into the prompt and stays alive:
+
+```shell
+$ q ${TORQHOME}/torq.q -proctype test -procname test1 -test /path/to/my/tests -debug
+```
+
+This will automatically load in the test code and all CSVs in that directory into a test TorQ process and run the tests in those files, then output the results to the screen.
+
+**Simple RDB Test**
+
+A similar scenario is when the test code is loaded into a 'production' TorQ process, ie. an RDB or a tickerplant, in order to test the functioning of that process. A simple RDB example can be seen here:
+
+```
+action,ms,bytes,lang,code,repeat,minver,comment
+true,0,0,q,`segmentedtickerplant~.rdb.tickerplanttypes,1,,"Check the TP type"
+true,0,0,q,all (null;{-11h=type x}) @' .rdb.subscribeto,1,,"Check RDB is subscribed to null symbol (all tables)"
+true,0,0,q,1~count key .rdb.subcsv,1,,"Check that the subcsv exists"
+```
+
+These tests simply check that certain process variables have been set properly, and these are triggered similarly to the previous set, except that an RDB process is used to load in the tests rather than a 'blank' test process:
```

Comment:
It might be worth mentioning here that by using the appropriate `proctype`, relevant config will be loaded in, as well as `${KDBCODE}/code/{procname}/` folder depending on config (`.proc.loadprocesscode` is the relevant config variable for that, I think - we have it turned on for most processes in TorQ by default I think).

This allows what I think is a fairly clean, test-friendly pattern for developing a TorQ process - placing majority of logic in q script(s) in `code/{procname}/` & then having a "main" `code/processes/{procname}.q` script that simply initialises the process (connects to discovery & any other procs required, does any subscriptions/recovery, sets timers etc.). That way, using a line like the one below, all the logic can be loaded without initialising the process, and functions can be easily tested in a standalone fashion.

`procname` generally isn't as important (can have procname specific config & code, but typically it's proctype), and can be set to something other than e.g. `rdb1` to allow tests to run without conflicting with an already running stack.

Reply:
I've added a note on this stuff, but are unit testing docs the best place to go into detail on it?

Reply:
What you have added looks good, other than this bit:

> which loads in the code in `${KDBCODE}/processes/rdb.q`

that's not true, for regular TorQ rdb that file is loaded by `-load` flag, with just the `-procname` it _won't_ be loaded (which is good - for unit testing, don't want to initialise the process fully & make connections to other TorQ processes etc.). It'll just be the relevant config & (usually) code from `${KDBCODE}/rdb` ("usually" because loading that code is config-dependent, but usually turned on for TorQ processes)